### DISCUSSION TITLE
(how) can I use convert function?

### DISCUSSION THREAD

**Discussion by alxgomz** [docker/buildx#3752] [Q&A]
Not sure whether that's a problem with me or a bug but here is what I'm trying to do:

I wrote a function in my bake file to compute a list of targets based on an whether an env var (`MY_
VERSION`) matches a list of `versions`:

```hcl
function "exclude_if_version" {
  params = [
    versions,
    inputlist,
    excludees
    ]
  result = sethaselement(convert(versions, list(string)),"${MY_VERSION}") ? setsubtract(inputlist, excludees) : inputlist
}
```

As you can see above, I'd like to deal with cases where someone using that function passes numbers (e.g. `[3, 4]`) or string representation of the version (`["3.0", "4.0"]`) in the `versions` list.
Though when running that simple code I get the error below:

```
Invalid expression; function "list" does not exist: undefined
```

I've tried with more simple types and get the same error (e.g. `convert("true", bool)`)

I would have expected that to work given the [bake Reference](https://docs.docker.com/build/bake/reference/#function) states:
>  A [set of general-purpose functions](https://github.com/docker/buildx/blob/master/bake/hclparser/stdlib.go) provided by [go-cty](https://github.com/zclconf/go-cty/tree/main/cty/function/stdlib) are available for use in HCL files

**Comment by wellkilo** [comment]
cty's convert function takes a **type constraint expression**, and current Buildx registers HCL's type-aware ConvertFunc. The error shown came from an older Buildx/Bake parser that treated list(string) as an ordinary call to an undefined function named list.

More importantly, a cty list is homogeneous before your function runs. A caller cannot pass [3, "4.0"] as one list and defer normalization to convert. Let HCL coerce the list at its boundary, or normalize values individually as strings.

The first fix is therefore to update Buildx. On a current release this is valid:

    convert(versions, list(string))
    convert("true", bool)

If you need compatibility with the old version, make the function contract a list of strings instead:

    function "exclude_if_version" {
      params = [versions, inputlist, excludees]
      result = contains(versions, MY_VERSION) ? setsubtract(inputlist, excludees) : inputlist
    }

Call it with string versions:

    exclude_if_version(["3", "4.0"], inputlist, excludees)

If callers have numeric values, normalize them when declaring/calling the function (for example use quoted versions or format("%v", value) where appropriate). A cty list is homogeneous, so a truly mixed list is not a useful input contract even after upgrading.

You can inspect the current registration in [bake/hclparser/stdlib.go](https://github.com/docker/buildx/blob/3046cc7bafa7a985c07870819bee3632d1749b66/bake/hclparser/stdlib.go#L57-L59), which uses typeexpr.ConvertFunc. The current generated stdlib docs also show convert("42", number). Thus, on current Buildx list(string) and bool are type expressions specifically inside convert; they are not standalone runtime functions.
